include/bitops.h

```diff
@@ -0,0 +1,224 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_BITOPS_H
+#define DOSBOX_BITOPS_H
+
+#include <cassert>
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+// An enum and functions to help with bit operations
+// Works with any unisigned integer type up to 32-bits
+// (Extending to 64-bits is currently not needed)
+//
+// Examples
+// ~~~~~~~~
+// uint8_t reg = 0;
+//
+// Setting:
+//   set_bits(reg, bit_0 | bit_1); // 0b0000'0011
+//   set_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//   set_all_bits(reg); // 0b1111'1111
+//
+// Toggling:
+//   toggle_bits(reg, bit_4 | bit_5 | bit_6 | bit_7); // 0b0000'1111
+//   toggle_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//   toggle_all_bits(reg); //  0b1111'0000
+//
+// Clearing:
+//   clear_bits(reg, bit_4 | bit_5); // 0b1100'0000
+//   clear_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//
+// Checking:
+//   are_bits_set(reg, bit_6); // true
+//   are_bits_set(reg, bit_7); // true
+//   are_bits_set(reg, bit_8); // ASSERT FAIL (out of range)
+//   are_bits_set(reg, bit_6 | bit_7); // true (both need to be set)
+//   are_bits_set(reg, bit_5 | bit_6); // false (both aren't set)
+//   are_any_bits_set(reg, bit_5 | bit_6); // true (bit 6 is)
+//   are_bits_cleared(reg, bit_0 | bit_1 | bit_2 | bit_3); // true
+//
+
+enum BITS {
+	bit_0 = 1 << 0,
+	bit_1 = 1 << 1,
+	bit_2 = 1 << 2,
+	bit_3 = 1 << 3,
+	bit_4 = 1 << 4,
+	bit_5 = 1 << 5,
+	bit_6 = 1 << 6,
+	bit_7 = 1 << 7,
+	bit_8 = 1 << 8,
+	bit_9 = 1 << 9,
+	bit_10 = 1 << 10,
+	bit_11 = 1 << 11,
+	bit_12 = 1 << 12,
+	bit_13 = 1 << 13,
+	bit_14 = 1 << 14,
+	bit_15 = 1 << 15,
+	bit_16 = 1 << 16,
+	bit_17 = 1 << 17,
+	bit_18 = 1 << 18,
+	bit_19 = 1 << 19,
+	bit_20 = 1 << 20,
+	bit_21 = 1 << 21,
+	bit_22 = 1 << 22,
+	bit_23 = 1 << 23,
+	bit_24 = 1 << 24,
+	bit_25 = 1 << 25,
+	bit_26 = 1 << 26,
+	bit_27 = 1 << 27,
+	bit_28 = 1 << 28,
+	bit_29 = 1 << 29,
+	bit_30 = 1 << 30,
+	bit_31 = 1 << 31
+};
+
+// Is the bitmask within the maximum value of the register?
+template <typename T>
+static constexpr void validate_bit_width([[maybe_unused]] const T reg, const int bits)
+{
+	// ensure the register is unsigned
+	static_assert(std::is_unsigned<T>::value, "register must be unsigned");
+
+	// Ensure the bits fall within the type size
+	assert(static_cast<uint64_t>(bits) > 0);
+	assert(static_cast<int64_t>(bits) <=
+	       static_cast<int64_t>(std::numeric_limits<T>::max()));
+}
+
+// Set the indicated bits
+template <typename T>
+constexpr T set_bits_const(const T reg, const int bits)
+{
+	validate_bit_width(reg, bits);
+	return reg | static_cast<T>(bits);
+}
+template <typename T>
+constexpr void set_bits(T &reg, const int bits)
+{
+	reg = set_bits_const(reg, bits);
+}
```

Comment:
Are template function overloads allowed? If so, might I suggest keeping the same name for both variants?

Reply:
They're allowed, but even for a fixed type (if we left templates out of this), the signature becomes ambiguous, unfortunately. 
So the only option is to use a different name for the read-only returning type versus the modifying type.
Any ideas how to name them better?

Reply:
How about using `const T &reg`?

Reply:
Will try 👍

Reply:
Unfortunately can't be differentiated:

### Overload w/ `const T &reg`

![2022-01-25_14-15](https://user-images.githubusercontent.com/1557255/151068923-6ff00c58-cbb9-4f68-8d96-e05d1eeacf8e.png)

![2022-01-25_14-16](https://user-images.githubusercontent.com/1557255/151068978-b2cb9c98-07da-4bed-a23d-625835f69c36.png)

---

### Overload w/ `const T reg`

![2022-01-25_14-17](https://user-images.githubusercontent.com/1557255/151069243-99526460-965b-4bea-bfe8-488b7d9f5307.png)

![2022-01-25_14-18](https://user-images.githubusercontent.com/1557255/151069240-0b7f02fd-a232-4d6f-9a07-8d071cf0d975.png)

Reply:
Interesting in the `const T &reg` case the compiler isn't smart enough to choose the correct overload. I suspect you could force the issue by casting `reg` to const when calling, but that isn't very ergonomic to use.

Hmm, not sure what to suggest now. Are the const versions really needed?

Reply:
They're needed to perform on-the-fly changes against existing variables that we don't want to modify.
Here's a real world example:

``` c++
if ((port_61_data ^ val) & 3)
	if ((port_61_data ^ val) & 1)
		TIMER_SetGate2(val&0x1);

We want to flip port_61_data's `val` bits and then check if either bit 0 or bit 1 are set.

We don't want to change port_61_data.

An almost rote conversion would be:

```c++
const auto p61_val_flipped = flip_bits_const(port_61_data, val);
if (are_any_bits_set(p61_val_flipped, bit_0 | bit_1))
	if (are_any_bits_set(p61_val_flipped, bit_0))
		TIMER_SetGate2(are_bits_set(val, bit_0));
```

Reply:
If we invert the question: are the modifying reference versions really needed?

In theory, they're not, but they really help make the code self-documenting. 

Here's an example from Spellicopter:

``` c++
static uint8_t read_p61(io_port_t, io_width_t)
{
	// Bit 4 must be toggled each request
	toggle_bits(port_61_data, bit_4);

	// On PC/AT systems, bit 5 sets the timer 2 output status
	if (is_machine(MCH_EGA | MCH_VGA))
		set_bits_to(port_61_data, bit_5, TIMER_GetOutput2());
	else
		// On XT systems always toggle bit 5 (Spellicopter CGA)
		toggle_bits(port_61_data, bit_5);

	return port_61_data;
}
```

But not returning a value, it's very clear that we're performing the changes on the register itself.

If we only had the const version, the code would become:


``` c++
static uint8_t read_p61(io_port_t, io_width_t)
{
	// Bit 4 must be toggled each request
	port_61_data = toggle_bits(port_61_data, bit_4);

	// On PC/AT systems, bit 5 sets the timer 2 output status
	if (is_machine(MCH_EGA | MCH_VGA))
		port_61_data = set_bits_to(port_61_data, bit_5, TIMER_GetOutput2());
	else
		// On XT systems always toggle bit 5 (Spellicopter CGA)
		port_61_data = toggle_bits(port_61_data, bit_5);

	return port_61_data;
}
```

Reply:
So I very much want both, as a user, but all for any simplifications we can make the names. 

Because these are so low-level, I think we can make them more terse as well; so I would be OK with short-hand notation (given it's something that, hopefully, we can put to use in many places).

Reply:
See updated version.